File: frontend/src/types/judge.ts
// TODO: Consider generating these types from the FastAPI OpenAPI spec
// using openapi-typescript once the backend is implemented.
//
// TODO: Judge types have significant structural differences from backend schemas.
// Backend JudgeConfigResponse uses flat fields (model, temperature, dimensions as
// list[dict[str, Any]]) while frontend models a panel of members with weights and
// typed dimensions. Needs a coordinated redesign.
// See https://github.com/narmaku/eval-studio/issues/106.

export type AggregationStrategy = 'majority_vote' | 'average' | 'max' | 'weighted_average';

export interface Judge {
  id: string;
  name: string;
  is_preset: boolean;
  panel: JudgePanelMember[];
  aggregation: AggregationStrategy;
  prompt_template: string;
  pass_threshold: number;
  dimensions: JudgeDimension[];
  created_at: string;
  updated_at: string;
}

export interface JudgePanelMember {
  model: string;
  temperature: number;
  weight: number;
}

export interface JudgeDimension {
  name: string;
  weight: number;
}

export interface JudgePreset {
  id: string;
  name: string;
  description: string;
  judge: Judge;
}

export interface CreateJudgeRequest {
  name: string;
  panel: JudgePanelMember[];
  aggregation: AggregationStrategy;
  prompt_template: string;
  pass_threshold: number;
  dimensions: JudgeDimension[];
}

File: backend/app/schemas/judge.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class JudgeConfigCreate(BaseModel):
    """Schema for creating a judge configuration."""

    name: str
    preset: str | None = None
    model: str | None = None
    temperature: float = 0.0
    prompt_template: str | None = None
    pass_threshold: float = 0.7
    dimensions: list[dict[str, Any]] | None = None
    aggregation: str | None = None


class JudgeConfigUpdate(BaseModel):
    """Schema for updating a judge configuration."""

    name: str | None = None
    model: str | None = None
    temperature: float | None = None
    prompt_template: str | None = None
    pass_threshold: float | None = None
    dimensions: list[dict[str, Any]] | None = None
    aggregation: str | None = None


class JudgeConfigResponse(BaseModel):
    """Schema for a judge configuration in API responses."""

    id: str
    name: str
    preset: str | None
    model: str | None
    temperature: float
    prompt_template: str | None
    pass_threshold: float
    dimensions: list[dict[str, Any]] | None
    aggregation: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
